### PR TITLE
✨ New API: OpenStackCluster.Spec.ExternalNetwork

### PR DESCRIPTION
**Pull request by EmilienM**
**What this PR does / why we need it**:

This PR converts `OpenStackCluster.Spec.ExternalNetworkID` to `OpenStackCluster.Spec.ExternalNetwork`.

The field `externalNetworkID` has been renamed to `externalNetwork` and isn't a string anymore but a reference to a `NetworkFilter` object.

```yaml
externalNetworkID: "e60f19e7-cb37-49f9-a2ee-0a1281f6e03e"
```

becomes

```yaml
externalNetwork:
  id: "e60f19e7-cb37-49f9-a2ee-0a1281f6e03e"
```

It is now possible to specify a `NetworkFilter` object to select the external network to use for the cluster. The `NetworkFilter` object allows to select the network by name, by ID or by tags.

```yaml
externalNetwork:
  name: "public"
```

If a network is provided, it'll be added to `OpenStackCluster.Status.ExternalNetwork`. If the network can't be found, an error will be returned.
If no network is provided, CAPO will try to find network marked "External" and add it to `OpenStackCluster.Status.ExternalNetwork`. If it can't find a network marked "External",
`OpenStackCluster.Status.ExternalNetwork` will be set to nil.
If more than one network is found, an error will be returned.

It is now possible for a user to specify that no external network should be used by setting `DisableExternalNetwork` to `true`:

```yaml
disableExternalNetwork: true
```

- [X] squashed commits
- [X] API conversion
- [X] includes documentation
- [X] adds unit tests

### CODE REVIEW

**Comment by EmilienM** [comment]
Ready for review, CI all green.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (EmilienM) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:19]: Typo here, will fix it

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (EmilienM) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:38]: Note for myself, the code doesn't allow to get the network by tags. Also I'll need to add a few validations to ensure we get one returned network.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (dulek) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:47]: This implies that when network is provided but isn't found, the CAPO will silently ignore the error. I believe that's wrong and if I read the code correctly it currently has that behavior.

**Review by dulek — CHANGES_REQUESTED** [review]

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (EmilienM) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:47]: yes exactly :(

File: api/v1alpha7/conversion_test.go
Comment (EmilienM) [inline api/v1alpha7/conversion_test.go:67]: I need to update this comment here and in v5 too, bad copy paste

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (EmilienM) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:47]: yeah I'm fine returning an error.

Comment (EmilienM) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:47]: Do you know why we strictly need an external network? If a cluster is deployed on a provider network and reachable from outside already, this should not be needed, right? (ie no needs of floating ips, etc).

Comment (dulek) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:47]: I'm pretty sure we have it for two reasons:

1. If we're creating a router, we need to point it to an external network to allow connectivity.
2. We need a network to create FIPs in (and later on to pass it to CCM so it creates FIPs there).

My main point here is that currently the code silently ignores when user specified a non-existing network. IMO it can just go over it when user haven't specified it, but this isn't really a good user experience if we don't inform users that their input was wrong.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (MaysaMacedo) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:6]: looks like this section is missing from the docs

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (EmilienM) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:47]: But:

1. Router creation is optional, right?
2. Why do we need FIPs? What about when deploying the cluster on a pre-existing network that is already routed (e.g. provider network)

Comment (EmilienM) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:47]: Ok we agreed offline that:
* external network is not strictly required
* if the user provides an external network we will change the behaviour to return a failure if that network wasn't found
* investigate how a user can specify that no external network is desired

File: pkg/cloud/services/networking/network.go
Comment (dulek) [inline pkg/cloud/services/networking/network.go:110]: Typo.

File: pkg/cloud/services/networking/network.go
Comment (dulek) [inline pkg/cloud/services/networking/network.go:109]: Neat trick. We could maybe name the var `isAutodetecting` or something to make it simpler to understand.

**Review by dulek — APPROVED** [review]
/lgtm

Looks good to me, minor comments inline, no need to spin the CI again just for them.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (dulek) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:19]: v1alpha8

File: api/v1alpha8/openstackcluster_types.go
Comment (MaysaMacedo) [inline api/v1alpha8/openstackcluster_types.go:75]: Could you add a comment describing the purpose of this field?

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (MaysaMacedo) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:60]: Might be worth detailing what "auto" means in the context of `externalNetworkType`

File: config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml
Comment (MaysaMacedo) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml:5264]: There is pattern for naming the API fields that mean to disable something, which I think is what this  does right?
might be worth to rename it to `disableExternalNetwork`.

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/api/v1alpha8/openstackcluster_types.go#L91

File: pkg/cloud/services/networking/network_test.go
Comment (MaysaMacedo) [inline pkg/cloud/services/networking/network_test.go:312]:
```suggestion
			name: "not reconcile external network when external network is disabled",
```

**Comment by MaysaMacedo** [comment]
lgtm besides the last nit I mentioned

File: api/v1alpha8/openstackcluster_types.go
Comment (stephenfin) [inline api/v1alpha8/openstackcluster_types.go:77]: Question: given `ExternalNetwork` is optional, couldn't we simply suggest that leaving this undefined/explicitly set to `null` indicates that the external network should be disabled?

*Later:* Okay, an unset `externalNetwork` value already carries a special meaning. I wonder if we can change that meaning. I discuss this later.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (stephenfin) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:29]:
```suggestion
The field `externalNetworkID` has been renamed to `externalNetwork` and isn't a string anymore but a reference to a `NetworkFilter` object.

The field `externalNetworkID` has been renamed to `externalNetwork` and is now a `NetworkFilter` object rather than a string ID.
The `NetworkFilter` object allows selection of a network by name, by ID or by tags.
```

to mirror what we're saying in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1796

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (stephenfin) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:51]: Ah, okay, this is the reason for the new `disableExternalNetwork` knob.

Given we're introducing the ability to filter external networks by name as well as a new API version, could we drop this feature and instead instead users specify an network (or omit it). This warrants a big warning in the migration docs since it'll be a larger (though hopefully easy to spot) change.

**Review by stephenfin — CHANGES_REQUESTED** [review]
I'm wondering if we can get rid of the auto-configuration of external network feature (anti-feature?). Does anyone rely on this?

File: api/v1alpha8/openstackcluster_types.go
Comment (EmilienM) [reply api/v1alpha8/openstackcluster_types.go:77]: > couldn't we simply suggest that leaving this undefined/explicitly set to null indicates that the external network should be disabled?

I thought about it too but we can't because to remain backward compatible, we have to support that users using the default (not set), CAPO will try to find an external network for them.
If we are ok breaking this then indeed we don't need `DisableExternalNetwork`.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (EmilienM) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:54]: I'm fine with it. Waiting for @mdbooth's feedback and if he agrees I can propose this change.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (stephenfin) [inline docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:29]:
```suggestion
```

I think you want to remove this line?

File: pkg/cloud/services/networking/network_test.go
Comment (mdbooth) [inline pkg/cloud/services/networking/network_test.go:345]: In this and the test above these should probably be nil.

File: pkg/cloud/services/networking/network_test.go
Comment (mdbooth) [inline pkg/cloud/services/networking/network_test.go:1]: Great tests, thanks!

File: api/v1alpha7/conversion_test.go
Comment (mdbooth) [inline api/v1alpha7/conversion_test.go:90]: As with the ServerGroup PR, this doesn't look right to me. Lets investigate.

File: api/v1alpha7/conversion.go
Comment (mdbooth) [inline api/v1alpha7/conversion.go:263]: Could we have some test coverage of these conversions? I think I'd probably test at the OpenStackCluster level as that's the most relevant unit here, which would also make it easy to expand to cover other conversions.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (mdbooth) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:53]: I don't have any hard data, but I suspect there are a bunch of folks who like the default. Removing it requires the user to be explicit about something that was previously implicit, so we wouldn't be able to convert it.

An alternative might be instead of `disableExternalNetwork` to have `useDefaultExternalNetwork`, i.e. you have to opt-in to the current default.

The disableExternalNetwork knob also allows folks with multiple external networks to not use any of them, which isn't currently possible.

I don't have a strong opinion, tbh. I'm ok with @EmilienM's version.

Soliciting opinions from @lentzi90 @mnaser @mkjpryor.

File: api/v1alpha7/conversion.go
Comment (EmilienM) [reply api/v1alpha7/conversion.go:263]: Just to make sure I understand what you suggest: adding unit tests that check that conversion produced an expected OpenStackCluster object. For example like it was done in v1alpha6 for PortOpts convert stuffs?

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (mkjpryor) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:53]: Personally, I quite like the fact that if there is exactly one external network then that gets used without further intervention.

I do love the fact that I will now be able to select my external network by Neutron tag now though, like I can with my cluster network.

**Comment by EmilienM** [comment]
/hold
I've added cherry-picks to fix conversion, until Matt's PR has merged.

File: docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md
Comment (lentzi90) [reply docs/book/src/topics/crd-changes/v1alpha7-to-v1alpha8.md:83]: I agree with @mkjpryor , this is convenient and nice

**Review by dulek — APPROVED** [review]
/lgtm from my side too.

**Comment by mdbooth** [comment]
I just pushed an update which fixes conversion of the new fields. The changes are:
* For OpenStackCluster we use the existing spec restorer, and add restoration of the lost fields in `restorev1alpha8ClusterSpec`
* We add a filter to the cluster spec restorer to fix the overlap with the bastion restorer.
* We change OpenStackCluster template to just have a single big restorer for the spec, as we don't need to worry about frequent updates to a template

**Comment by mdbooth** [comment]
/approve

I'm happy for this to merge once somebody else has looked it over.

File: pkg/cloud/services/networking/network.go
Comment (MaysaMacedo) [inline pkg/cloud/services/networking/network.go:77]: this doc looks outdated

**Pull request merged**
